### PR TITLE
Feature 154 front end tracker

### LINKED ISSUE #154: Front end tracker
Develop the tracker class for the front end.

The tracker should have the following features:

- take in an image stream
- detect, describe, and match keypoints (templated on classes already in libwave)
- Assign IDs to each good landmark.
- Match for either both monocular or multi-cam clusters over time
- visualize feature tracks
- return a populated measurement container with landmark IDs

### PR DESCRIPTION
**Pull request by navganti**
Resolves #154. Contains methods to add images, extract tracks, and also draw tracks onto images. Also includes an "offline" tracker, which can track features over a sequence of images. Examples of feature tracks are attached, using the FASTDetector, BRISKDescriptor, and BruteForceMatcher.

There are also some minor changes to some other modules.

#### Pre-Merge Checklist:
- [x] Code is documented in doxygen format
- [x] Code has automated tests
- [x] Zero compiler warnings
- [x] Formatted with `clang-format`

### CODE REVIEW

File: wave_vision/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 2.8.3)
 PROJECT(wave_vision)
+# SET(CMAKE_BUILD_TYPE RELEASE)
```
Comment (chutsu) [inline wave_vision/CMakeLists.txt]: Why is this here?

Comment (navganti) [inline wave_vision/CMakeLists.txt]: Oops. I had that for testing the speed of some algorithms during testing. Thanks!

**Comment by Jebediah** [comment]
cool. What are the arrows on the pictures?

**Comment by navganti** [comment]
@Jebediah the arrows represent the pixel locations of how the features have moved over time. Depending on how many images the feature has been seen in, the tracks will be longer.

There are clearly some outliers. My thought is that this is due to the use of a pretty simple detector/descriptor, and a more robust method may remove them.

**Comment by chutsu** [comment]
Its pointing to where the feature was last seen from, its a bit backwards but ¯\_(ツ)_/¯

**Comment by chutsu** [comment]
Damnit nav beat me

**Comment by navganti** [comment]
@chutsu I'm not sure I agree that it's backwards? It's pointing to where the feature currently is, from where it came from. IMO it doesn't really make sense to point the other way, since it's illustrating the motion of the feature over time.

**Comment by chutsu** [comment]
Already updated my comment

**Comment by navganti** [comment]
@chutsu do you think it should be the other way?

**Comment by Jebediah** [comment]
So any of the arrows that kind of go across the middle of the image are outliers?

**Comment by chutsu** [comment]
Nope, I just had a brain fart moment.

**Comment by navganti** [comment]
@chutsu now my comments look dumb :stuck_out_tongue:

**Comment by navganti** [comment]
@Jebediah Yup. Since the car is only moving forward in these scenes, it doesn't really make sense that a feature can be on one side of the image in one timestep, and then jump across in the next. The best tracks are the traffic lights in the top right - most other features should follow a similar path.

**Comment by Jebediah** [comment]
Have you tried it on non-blurry images?

**Comment by navganti** [comment]
@Jebediah no, not yet. I definitely should...I'll test it out with a few Kitti images.

File: wave_vision/include/wave/vision/utils.hpp [hunk 2]
```diff
@@ -43,20 +43,54 @@ Vec2 focal_length(double hfov,
...
+cv::Point2f convertKeypoint(const Vec2 &keypoint);
+
+/** Convert a vector of cv::KeyPoint to a vector of Vec2
  *
  * @param keypoints input keypoints
  * @param vec_keypoints converted keypoints
  */
-void convertKeypoints(const std::vector<cv::KeyPoint> &keypoints,
-                      std::vector<Vec2> &vec_keypoints);
+std::vector<Vec2> convertKeypoints(const std::vector<cv::KeyPoint> &keypoints);
```
Comment (leokoppel) [inline wave_vision/include/wave/vision/utils.hpp]: These vector conversion functions may be excessive if you're not using them a whole lot (and it looks like they are not used).

You can use `std::transform` apply a function to a whole vector if needed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        feature_tracks = {};
+    }
+
+    return feature_tracks;
+}
+
+template <typename TDetector, typename TDescriptor, typename TMatcher>
+void Tracker<TDetector, TDescriptor, TMatcher>::addImage(const cv::Mat &image) {
+    // Get time for keypoint registration
+    std::chrono::steady_clock::time_point current_time = clock.now();
```
Comment (leokoppel) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: This is probably a `todo`, using times provided from outside. Maybe time should be an argument to `addImage`.
It doesn't make sense to get the current system time in offline tracking.

File: wave_vision/include/wave/vision/tracker/tracker.hpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+namespace wave {
+/** @addtogroup vision
+ *  @{ */
+
+class FeatureTrack {
+ public:
+    // Constructor
+    FeatureTrack() {}
+
+    // Initializer list constructor
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: Sorry, will remove. I like to think I'm not writing redundant comments anymore but apparently a few still make it through :stuck_out_tongue:

File: wave_vision/include/wave/vision/tracker/tracker.hpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+#include "wave/containers/landmark_measurement.hpp"
+#include "wave/containers/landmark_measurement_container.hpp"
+#include "wave/utils/utils.hpp"
+#include "wave/vision/utils.hpp"
+
+namespace wave {
+/** @addtogroup vision
+ *  @{ */
+
+class FeatureTrack {
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: They are. This was the format @chutsu and I talked about. In particular the first/last images were parameters that he wanted. @chutsu, can you look at the LandmarkID class and see if that's fine as is? It would be nice to remove this class

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp (hunk 3 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: I'm not sure if this makes sense to get the time here as well. Actually, I just realized this should read `this->clock.now();`.

But does that even make sense in an online setting? This is currently set up as a member clock. I think whether it's online or offline, it should take a time from elsewhere.

But for offline, times maybe as a whole don't make sense, which you've mentioned below.

File: wave_vision/include/wave/vision/tracker/tracker.hpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    std::vector<std::vector<FeatureTrack>> offlineTracker(
+      const std::vector<cv::Mat> &image_sequence);
+
+ private:
+    /** Generate a new ID for each newly detected feature.
+     *
+     * @return the assigned ID.
+     */
+    size_t generateFeatureID() const {
+        static size_t id = 0;
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: I think that's fine. That's a good point, I wasn't aware of that. The actual ID assigned should not really matter as long as its unique...do you think it's okay as is?

PS thanks, but I stole this use of static from your MSCKF @leokoppel :smile:

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+  const std::vector<cv::DMatch> &matches) {
+    // Maps current keypoint indices to IDs
+    std::map<int, size_t> curr_ids;
+
+    // TODO will need way to specify camera sensor ID.
+    int sensor_id = 0;
+
+    for (const auto &m : matches) {
+        // Check to see if ID has already been assigned to keypoint
+        if (this->prev_ids.find(m.queryIdx) != this->prev_ids.end()) {
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+            this->prev_ids[m.queryIdx] = this->generateFeatureID();
+            curr_ids[m.trainIdx] = this->prev_ids.at(m.queryIdx);
+
+            // Since keypoint was not a match before, need to add previous and
+            // current points to measurement container
+            Vec2 prev_landmark = convertKeypoint(this->prev_kp.at(m.queryIdx));
+            Vec2 curr_landmark = convertKeypoint(curr_kp.at(m.trainIdx));
+
+            // Find previous and current times from lookup table
+            std::chrono::steady_clock::time_point prev_time =
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: Fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+    return curr_ids;
+}
+
+// Public Functions
+template <typename TDetector, typename TDescriptor, typename TMatcher>
+std::vector<FeatureTrack> Tracker<TDetector, TDescriptor, TMatcher>::getTracks(
+  const size_t &img_num) const {
+    std::vector<FeatureTrack> feature_tracks;
+    std::vector<size_t> landmark_ids;
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp (hunk 7 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp (hunk 8 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp (hunk 9 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+            size_t first_image;
+
+            // Last image feature was seen in must be image requested.
+            size_t last_image = img_num;
+
+            // Iterators for time map and landmark measurements
+            std::map<size_t,
+                     std::chrono::steady_clock::time_point>::const_iterator
+              times;
+            std::vector<LandmarkMeasurement<int>>::const_iterator first =
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+                    first_image = times->first;
+                }
+            }
+
+            // Emplace new feature track back into vector
+            feature_tracks.emplace_back(l, tracks, first_image, last_image);
+        }
+    } else {
+        // img_num is zero, and there are no tracks.
+        feature_tracks = {};
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        // Detect, describe, and match keypoints
+        this->detectAndCompute(image, curr_kp, curr_desc);
+        matches = this->matcher.matchDescriptors(
+          this->prev_desc, curr_desc, this->prev_kp, curr_kp);
+
+        // Register keypoints with IDs, and store Landmarks in container
+        curr_ids = this->registerKeypoints(curr_kp, matches);
+
+        // Set previous ID map to be the current one, and reset
+        this->prev_ids = curr_ids;
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: Fixed. Thanks!

File: wave_vision/include/wave/vision/tracker/tracker.hpp (hunk 5 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: So after talking with @chutsu offline, I think we'd like to keep the FeatureTrack class. This is primarily because of the first/last_image parameters. This may tie in to the other comment regarding times/image counts for the measurement container...we should discuss this offline @leokoppel.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp (hunk 3 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: Okay yes. Will add as an input to the function.

Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/tracker.hpp [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    /** The id of the feature */
+    size_t id;
+
+    /** All measurements of the feature */
+    std::vector<LandmarkMeasurement<int>> measurements;
+
+    /** The first image the feature is seen in */
+    size_t first_image;
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: I don't know if I want to define the type of the `first_image` to be `LandmarkId`. I think that could be a little confusing, no? Like if there was another typedef for `ImageId` or something that may make sense.

Comment (leokoppel) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: Yes you're right, I was mistaken.  (And that's why it would be confusing to use size_t for everything) 
Yes add an alias, especially since you're adding a field to LandmarkMeasurement

File: wave_vision/include/wave/vision/tracker/tracker.hpp [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    Tracker(TDetector detector, TDescriptor descriptor, TMatcher matcher)
+        : detector(detector), descriptor(descriptor), matcher(matcher) {}
+
+    /** Destructor */
+    ~Tracker() = default;
+
+    /** Get the tracks of all features in the requested image from the sequence.
+     *
+     * @param img_num the number of the image to obtain tracks from
+     * @return tracks corresponding to all detected landmarks in the image.
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: fixed.

File: wave_vision/include/wave/vision/tracker/tracker.hpp (hunk 13 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: Sure thing. I'll define that in the LandmarkMeasurement file once I get there.

File: wave_vision/include/wave/vision/tracker/tracker.hpp [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+                          std::vector<cv::KeyPoint> &keypoints,
+                          cv::Mat &descriptor);
+
+    /** Register the current time with the current img_count
+     *
+     * @param current_time the time at which this image was received
+     */
+    void timestampImage(
+      const std::chrono::steady_clock::time_point &current_time) {
+        this->img_times[this->img_count] = current_time;
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: Good catch. Fixed.

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp [hunk 16]
```diff
@@ -327,4 +347,46 @@ TEST(BFTests, DISABLED_LMEDS) {
 
     cv::waitKey(0);
 }
+
+TEST(BFTests, DISABLED_SequenceMatches) {
```
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: Yeah, in my other classes I only have 1 skipped test or so, but this one is just so image dependent. @leokoppel do you have any tips for testing this stuff a bit better? While trying to write tests I was struggling a bit with ensuring everything is tested properly. It's a bit tougher when visual results are pretty key for confirming operation.,

File: wave_vision/include/wave/vision/utils.hpp (hunk 2 above)
Comment (navganti) [inline wave_vision/include/wave/vision/utils.hpp]: I'm using one of them I think. I felt it was just useful to have for future use - these are the three main types a user may encounter in terms of keypoints.

File: wave_vision/include/wave/vision/tracker/tracker.hpp (hunk 15 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: Ah turns out they're not _quite_ equal since C++ is zero indexed. It's still unnecessary though, and was more confusing.

File: wave_vision/tests/descriptor_tests/brisk_tests.cpp [hunk 17]
```diff
@@ -164,9 +165,17 @@ TEST(BRISKTests, DISABLED_ComputeDescriptors) {
     FASTDetector fast;
     BRISKDescriptor brisk;
 
+    auto start = std::chrono::steady_clock::now();
+
     keypoints = fast.detectFeatures(image);
     descriptors = brisk.extractDescriptors(image, keypoints);
 
+    auto extract_time = std::chrono::duration_cast<std::chrono::milliseconds>(
```
Comment (navganti) [inline wave_vision/tests/descriptor_tests/brisk_tests.cpp]: This was just to get an idea of how fast things are running, but the results don't really make sense tbh. Removing this.

File: wave_vision/include/wave/vision/tracker/impl/tracker.hpp [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+            // Create LandmarkMeasurement, and add to
+            // LandmarkMeasurementContainer
+            LandmarkMeasurement<int> measurement(
+              this->img_times.at(this->img_count),
+              sensor_id,
+              curr_ids.at(m.trainIdx),
+              landmark);
+            this->landmarks.insert(measurement);
+        } else {
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/impl/tracker.hpp]: I think with the latest changes it's looking a bit cleaner. Let me know what you think @leokoppel

File: wave_vision/include/wave/vision/tracker/tracker.hpp (hunk 5 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: Okay, I have added the image number to the LandmarkMeasurements. The first and last images can be directly extracted from the vector of feature tracks pretty easily.

File: wave_vision/include/wave/vision/tracker/tracker.hpp [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+                 std::vector<LandmarkMeasurement<int>> measurements,
+                 size_t first_image,
+                 size_t last_image)
+        : id(id),
+          measurements(measurements),
+          first_image(first_image),
+          last_image(last_image) {}
+
+    // Destructor
+    ~FeatureTrack() = default;
```
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: done

File: wave_vision/include/wave/vision/tracker/tracker.hpp (hunk 13 above)
Comment (navganti) [inline wave_vision/include/wave/vision/tracker/tracker.hpp]: fixed.

File: wave_vision/src/utils.cpp [hunk 20]
```diff
@@ -27,15 +27,74 @@ void projection_matrix(const Mat3 &K, const Mat3 &R, const Vec3 &t, MatX &P) {
...
+    std::vector<cv::Point2f> cv_keypoints;
+
+    for (const auto &k : keypoints) {
+        cv_keypoints.emplace_back((float) k(0), (float) k(1));
+    }
+
+    return cv_keypoints;
+}
+
+std::vector<cv::Mat> readImageSequence(const std::string &path) {
```
Comment (navganti) [inline wave_vision/src/utils.cpp]: Mmm good point. I still think this could be a useful utility. Will revisit in the future :slightly_smiling_face:

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp (hunk 16 above)
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: Okay, fixed. Created `wave_vision_viz_tests`, and all visualization tests have been moved into that directory. As more modules are developed they can be added there.

**Comment by navganti** [comment]
@leokoppel @chutsu @Jebediah Here are the results with some non-blurry images (just picked a KITTI sequence). The car dips up and down quite a bit, so the feature tracks are not nearly as nice as the Ottawa set. That being said, there are very few outliers.
[kitti_tracks.zip](https://github.com/wavelab/libwave/files/1163628/kitti_tracks.zip)


[kitti_tracks.zip](https://github.com/wavelab/libwave/files/1163623/kitti_tracks.zip)
